### PR TITLE
New resource/ data source `azurerm_postgresql_flexible_server`

### PR DESCRIPTION
**Pull request by yupwei68**
Fix: https://github.com/terraform-providers/terraform-provider-azurerm/issues/8654
  https://github.com/terraform-providers/terraform-provider-azurerm/issues/10455

=== RUN   TestAccDataSourcePostgresqlflexibleServer_basic
=== PAUSE TestAccDataSourcePostgresqlflexibleServer_basic
=== CONT  TestAccDataSourcePostgresqlflexibleServer_basic
--- PASS: TestAccDataSourcePostgresqlflexibleServer_basic (485.30s)

=== RUN   TestAccPostgresqlflexibleServer_basic
=== PAUSE TestAccPostgresqlflexibleServer_basic
=== CONT  TestAccPostgresqlflexibleServer_basic
--- PASS: TestAccPostgresqlflexibleServer_basic (744.14s)
=== RUN   TestAccPostgresqlflexibleServer_requiresImport
=== PAUSE TestAccPostgresqlflexibleServer_requiresImport
=== CONT  TestAccPostgresqlflexibleServer_requiresImport
--- PASS: TestAccPostgresqlflexibleServer_requiresImport (800.16s)
=== RUN   TestAccPostgresqlflexibleServer_complete
=== PAUSE TestAccPostgresqlflexibleServer_complete
=== CONT  TestAccPostgresqlflexibleServer_complete
--- PASS: TestAccPostgresqlflexibleServer_complete (1620.08s)
=== RUN   TestAccPostgresqlflexibleServer_completeUpdate
=== PAUSE TestAccPostgresqlflexibleServer_completeUpdate
=== CONT  TestAccPostgresqlflexibleServer_completeUpdate
--- PASS: TestAccPostgresqlflexibleServer_completeUpdate (3408.08s)
=== RUN   TestAccPostgresqlflexibleServer_updateMaintenanceWindow
=== PAUSE TestAccPostgresqlflexibleServer_updateMaintenanceWindow
=== CONT  TestAccPostgresqlflexibleServer_updateMaintenanceWindow
--- PASS: TestAccPostgresqlflexibleServer_updateMaintenanceWindow (1124.51s)
=== RUN   TestAccPostgresqlflexibleServer_updateSku
=== PAUSE TestAccPostgresqlflexibleServer_updateSku
=== CONT  TestAccPostgresqlflexibleServer_updateSku
--- PASS: TestAccPostgresqlflexibleServer_updateSku (1472.27s)
=== RUN   TestAccPostgresqlflexibleServer_pitr
=== PAUSE TestAccPostgresqlflexibleServer_pitr
=== CONT  TestAccPostgresqlflexibleServer_pitr
--- PASS: TestAccPostgresqlflexibleServer_pitr (2484.59s)

### CODE REVIEW

File: azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+						"type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"sku": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go:68]: is there a reason this isn't sku_name ?

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+			},
+
+			"administrator_login_password": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Sensitive:    true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"sku": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:70]: again, is there a reason we are not using sku_name for this?

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+					string(postgresqlflexibleservers.Default),
+					string(postgresqlflexibleservers.PointInTimeRestore),
+				}, false),
+			},
+
+			"delegated_subnet_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: azure.ValidateResourceID,
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:141]: should we use a subnet id validater?

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+	client := meta.(*clients.Client).Postgres.FlexibleServersClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	id := parse.NewFlexibleServerID(subscriptionId, resourceGroup, name).ID()
+
+	existing, err := client.Get(ctx, resourceGroup, name)
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:270]: same as above
```suggestion
	existing, err := client.Get(ctx, id.ResourceGroup, id.Name)
```

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+
+			"administrator_login": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"administrator_login_password": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:63]: i think login is redundant here?
```suggestion
			"administrator_password": {
```

File: website/docs/r/postgresql_flexible_server.html.markdown [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+* `tier` - (Required) The SKU tier for the PostgreSQL Flexible Server. Possible values are `Burstable`, `GeneralPurpose`, or `MemoryOptimized`.
+
+## Attributes Reference
+
+In addition to the Arguments listed above - the following Attributes are exported: 
+
+* `id` - The ID of the PostgreSQL Flexible Server.
+
+* `byok_enforcement` - The status showing whether the data encryption is enabled with a customer-managed key.
```
Comment (katbyte) [inline website/docs/r/postgresql_flexible_server.html.markdown:136]: why is this byok when its talking about cmk? should it be
```suggestion
* `cmk_enabled` - The status showing whether the data encryption is enabled with a customer-managed key.
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @yupwei68 - i've given this a quick review and left some comments inline

File: azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go (hunk 1 above)
Comment (yupwei68) [reply azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go:68]: There is a `name` and `tier` required in the block `sku`. May this be wield to have a `name` in block `sku_name`?

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go (hunk 2 above)
Comment (yupwei68) [reply azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:70]: same above?

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go (hunk 4 above)
Comment (yupwei68) [reply azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:270]: id is a string variable?

**Comment by yupwei68** [comment]
Hi kt, thanks for your comments. Changes have been pushed. Please continue reviewing.

File: azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go (hunk 1 above)
Comment (katbyte) [reply azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go:68]: We changed postgres to sku_name and it has a tier so maybe we should make this the same?

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go (hunk 4 above)
Comment (katbyte) [reply azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:271]: if you use `id := parse.NewFlexibleServerID(subscriptionId, resourceGroup, name)` and then id.ID() when you need to full id string you can use the structure?

**Comment by katbyte** [comment]
addressed the two outstanding comments

**Comment by yupwei68** [comment]
Thanks kt, Corresponding changes have been pushed. Tests have passed. Please continue reviewing.

**Comment by dhirschfeld** [comment]
Looks like this missed the boat for 2.55. Would love to see it in the next one!

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 7]
```diff
@@ -0,0 +1,636 @@
...
+			},
+
+			"sku_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validate.FlexibleServerSkuName,
+			},
+
+			"sku_tier": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:78]: again, why are these still separate and not combined to be consistent with the other postgres resources?

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 8]
```diff
@@ -0,0 +1,636 @@
...
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(postgresqlflexibleservers.OneOne),
+					string(postgresqlflexibleservers.OneTwo),
+				}, false),
+			},
+
+			"availability_zone": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:107]: how come this is not just "zones" like most resources? could we make this consistent with the majority of resources? there are shared functions like `azure.SchemaZones(),` we can use

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+
+						"tenant_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"point_in_time_utc": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:166]: should this be
```suggestion
			"point_in_time_restore_time_in_utc": {
```

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 10]
```diff
@@ -0,0 +1,636 @@
...
+			},
+
+			"source_server_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.FlexibleServerID,
+			},
+
+			"ha_enabled": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:180]:
```suggestion
			"high_availiblity_enabled": {
```

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 11]
```diff
@@ -0,0 +1,636 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"fqdn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"ha_state": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:233]: here

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 12]
```diff
@@ -0,0 +1,636 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"public_network_access": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:238]:
```suggestion
			"public_network_access_enabled": {
```

File: azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+				Computed: true,
+			},
+
+			"version": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tags.SchemaDataSource(),
+		},
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go:84]: could we add all of the values from the resource here:?

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by yupwei68** [comment]
Thanks for your comments, kt. All tests have passed. Please continue reviewing when the linter is fixed.

File: azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+	client := meta.(*clients.Client).Postgres.FlexibleServersClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	id := parse.NewFlexibleServerID(subscriptionId, resourceGroup, name).ID()
+
+	resp, err := client.Get(ctx, resourceGroup, name)
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_data_source.go:107]: Again, can we not make id a string, and use the id struct throughout the function
```suggestion
	id := parse.NewFlexibleServerID(subscriptionId, resourceGroup, name).ID()

	resp, err := client.Get(ctx, id.ResourceGroup, id.Name)
```

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(postgresqlflexibleservers.OneOne),
+					string(postgresqlflexibleservers.OneTwo),
+				}, false),
+			},
+
+			"zone": {
```
Comment (katbyte) [inline azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:97]: I think there is a shared schema function you can use here?

**Review by katbyte — CHANGES_REQUESTED** [review]

File: azurerm/internal/services/postgres/postgresql_flexible_server_resource.go (hunk 15 above)
Comment (yupwei68) [reply azurerm/internal/services/postgres/postgresql_flexible_server_resource.go:97]: The shared schema is a list of string, while here it's a string, shall we change its type to use the shared schema?

Comment (katbyte) [reply azurerm/internal/services/postgres/postgresql_flexible_server_resource.go]: AH good point! we only have a couple `zone` - this is good

**Review by katbyte — APPROVED** [review]
Thanks @yupwei68 - LGTM 👍

**Pull request merged**